src/platform/console/terminal.cpp

```diff
@@ -58,11 +58,6 @@ bool mp::Terminal::is_live() const
 
 std::string mp::Terminal::read_all_cin()
 {
-    if (!cin_is_live())
-    {
-        throw std::runtime_error("cannot read from stdin without a TTY");
-    }
-
```

Comment:
The windows terminal sets the mode of `stdin` to binary before calling `read_all_cin()`. I am not sure what the intended effects are, but are we sure they're maintained outside of a live tty?

Reply:
Looking at [the docs for it](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setmode) I can't see any exceptions. It just sets the mode on the file descriptor.

Reply:
Yes, but to what end? Weren't there problems with this sort of stuff in the past? Like blocking exec commands and the like? This affects newlines, so could that be a cause? Could Gerry have put this in to fix something there?

Side note:

> If you apply _setmode to the file descriptor for a stream, call _setmode before you perform any input or output operations on the stream.

I wonder what happens if we don't, I don't think we ensure it.

Reply:
> Yes, but to what end? Weren't there problems with this sort of stuff in the past? Like blocking exec commands and the like? This affects newlines, so could that be a cause?

The problems you refer to were around the metrics question being asked, which should never be shown if we're not on a tty, or when there is anything to be read from stdin, because how do we differentiate between cloud-init being read and the answer to the question. Reading stdin and interactivity are conflicting. 

`exec` commands never reach `read_all_cin()`, it's only the cloud-init input handling that does.

>  Could Gerry have put this in to fix something there?

I can only guess that this was a wrong assumption that people will only feed it in from TTYs…

> Side note:
> 
> > If you apply _setmode to the file descriptor for a stream, call _setmode before you perform any input or output operations on the stream.
> 
> I wonder what happens if we don't, I don't think we ensure it.

One may wonder ;) You could also argue that for the YAML we _should_ be translating.  But it's out of scope for this PR.